File: haskell/acronym/src/Acronym.hs
module Acronym (abbreviate) where

import Data.Char (isSpace, isLetter, isLower, isUpper, toUpper)

abbreviate :: String -> String
abbreviate = concatMap initials . split

split :: String -> [String]
split [] = []
split (x:xs)
    | isSeparator x = split xs
    | otherwise     = word : split rest
    where (word, rest) = break isSeparator (x:xs)

isSeparator :: Char -> Bool
isSeparator c = isSpace c || c == '-'

initials :: String -> String
initials "" = ""
initials w
    | all isUpper letters   = [head w]
    | all isLower letters   = [toUpper (head w)]
    | otherwise             = filter isUpper w
    where letters = filter isLetter w

File: go/acronym/acronym.go
// Package acronym converts a phrase to its acronym
package acronym

import (
	"bytes"
	"strings"
)

// Abbreviate converts a phrase to its acronym
func Abbreviate(s string) string {

	var buffer bytes.Buffer

	for _, word := range strings.Split(strings.Replace(s, "-", " ", -1), " ") {
		buffer.WriteString(strings.ToUpper(string([]rune(word)[0])))
	}

	return buffer.String()
}
